### PR TITLE
Implement versioning custom action.

### PR DESCRIPTION
**Pull request by ylangisc**

### CODE REVIEW

**Review by dkocher — CHANGES_REQUESTED** [review]
Add integration test.

File: ctera/src/main/java/ch/cyberduck/core/ctera/CteraCustomActionVersioning.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    private final CteraSession session;
+    private final Path file;
+
+    public CteraCustomActionVersioning(final CteraSession session, final Path file) {
+        this.session = session;
+        this.file = file;
+    }
+
+    public void run() throws BackgroundException {
+        final String device = this.getDeviceName();
```
[the remark this answers is not in the document]
Comment (ylangisc) [reply ctera/src/main/java/ch/cyberduck/core/ctera/CteraCustomActionVersioning.java]: Yes, would be better.

File: ctera/src/main/java/ch/cyberduck/core/ctera/CteraCustomActionVersioning.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                IOUtils.write(content, out, StandardCharsets.UTF_8);
+            }
+            return file;
+        }
+        catch(IOException e) {
+            throw new DefaultIOExceptionMappingService().map("Error writing temporary file", e);
+        }
+    }
+
+    private String getDeviceName() throws BackgroundException {
```
[the remark this answers is not in the document]
Comment (ylangisc) [reply ctera/src/main/java/ch/cyberduck/core/ctera/CteraCustomActionVersioning.java:115]: Done.

File: ctera/src/main/java/ch/cyberduck/core/ctera/CteraCustomActionVersioning.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+                                    "</script> </head> <body onload = \"_onload()\">\n" +
+                                    "<form id=\"sForm\" method=\"POST\" action=\"%s/ServicesPortal/sso\">\n" +
+                                    "<input type=\"hidden\" name=\"targeturi\" value=\"?GUI_openFmFolder=%s&GUI_fmVersions=true\"/>\n" +
+                                    "<input type=\"hidden\" name=\"ctera_ticket\" value=\"%s\"/>\n" +
+                                    "</form> </body> </html>",
+                            new HostUrlProvider().withUsername(false).get(session.getHost()),
+                            URLEncoder.encode(file.getAbsolute(), StandardCharsets.UTF_8.name()),
+                            token);
+            final Local file = TemporaryFileServiceFactory.get().create(String.format("%s.html", new AlphanumericRandomStringService().random()));
+            new DefaultLocalTouchFeature().touch(file);
```
[the remark this answers is not in the document]
Comment (ylangisc) [reply ctera/src/main/java/ch/cyberduck/core/ctera/CteraCustomActionVersioning.java:89]: Yes.

File: ctera/src/main/java/ch/cyberduck/core/ctera/CteraCustomActionVersioning.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                        return mapper.readValue(entity.getContent(), Device[].class);
+                    }
+                });
+                for(Device device : devices) {
+                    if(device.uid.equals(id)) {
+                        return device.name;
+                    }
+                }
+            }
+            throw new InteroperabilityException("No token or device found");
```
Comment (dkocher) [inline ctera/src/main/java/ch/cyberduck/core/ctera/CteraCustomActionVersioning.java:135]: Throw _No token found in keychain_ when keychain returns null and _No device found_ when device array is empty.

File: ctera/src/test/java/ch/cyberduck/core/ctera/CteraCustomActionVersioningTest.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+@Category(IntegrationTest.class)
+public class CteraCustomActionVersioningTest {
+
+    @Test
+    public void testSessionToken() throws Exception {
+        final Host host = new Host(new CteraProtocol(), "mountainduck.ctera.me", new Credentials(
```
Comment (dkocher) [inline ctera/src/test/java/ch/cyberduck/core/ctera/CteraCustomActionVersioningTest.java:44]: Extract abstract test class like in other protocols for connection setup.

**Review by dkocher — CHANGES_REQUESTED** [review]

**Review by dkocher — APPROVED** [review]

**Pull request merged**